src/middleware/socket.test.js

```diff
@@ -0,0 +1,242 @@
+		it('all clients should connect to websockets', async () => {
+			const client = await createClients(1, 1);
+
+			expect(client.teacher[0].isConnected).to.be.true;
+			expect(client.student[0].isConnected).to.be.true;
+
+			client.teacher[0].close();
+			client.student[0].close();
+		});
+	});
+
+	describe('communication', () => {
+		it('client should send and receive data', async () => {
+			/*= =============  Client Setup  ============== */
+			const client = await createClients(1, 1);
+			/*= ========  End of Client Setup  =========== */
+			const course = await getCourseArray(1);
+
+			const patch = {
+				title: 'F=MA',
+			};
+
+			await client.teacher[0].joinChannel(course[0].courseId, course[0].lessonId);
+			await client.student[0].joinChannel(course[0].courseId, course[0].lessonId);
+
+			await new Promise(((resolve, reject) => {
+				client.student[0].on('course/:courseId/lessons patched', (data) => {
+					expect(data).to.include(patch);
+					resolve();
+				});
+
+				client.teacher[0].emit(
+					'patch',
+					`course/${course[0].courseId}/lessons`,
+					course[0].lessonId,
+					patch,
+				);
+
+				setTimeout(() => {
+					reject(new Error('Client not receiving message in time! (>4000ms)'));
+				}, 4000);
+			}));
+
+			patch.title = 'E=MC²';
+
+			await new Promise(((resolve, reject) => {
+				client.teacher[0].on('course/:courseId/lessons patched', (data) => {
+					expect(data).to.include(patch);
+					resolve();
+				});
+
+				client.student[0].emit(
+					'patch',
+					`course/${course[0].courseId}/lessons`,
+					course[0].lessonId,
+					patch,
+				);
+
+				setTimeout(() => {
+					reject(new Error('Client not receiving message in time! (>4000ms)'));
+				}, 4000);
+			}));
+
+			client.teacher[0].close();
+			client.student[0].close();
+		});
+
+		it('client should send and receive data only in joined Channels', async () => {
+			/*= =============  Client Setup  ============== */
+			const client = await createClients(2, 2);
+			/*= ========  End of Client Setup  =========== */
+			const course = await getCourseArray(2);
+
+			const patch1 = {
+				title: 'F=MA',
+			};
+			const patch2 = {
+				title: 'E=MC²',
+			};
+
+			await client.teacher[0].joinChannel(course[0].courseId, course[0].lessonId);
+			await client.student[0].joinChannel(course[0].courseId, course[0].lessonId);
+			await client.student[1].joinChannel(course[1].courseId, course[1].lessonId);
+			await client.teacher[1].joinChannel(course[1].courseId, course[1].lessonId);
+
+
+			await new Promise(((resolve) => {
+				client.student[1].on('course/:courseId/lessons patched', (data) => {
+					expect(data).to.not.include(patch1);
+				});
+				client.teacher[1].on('course/:courseId/lessons patched', (data) => {
+					expect(data).to.not.include(patch1);
+				});
+
+				client.teacher[0].emit(
+					'patch',
+					`course/${course[0].courseId}/lessons`,
+					course[0].lessonId,
+					patch1,
+				);
+
+				setTimeout(() => {
```

Comment:
emit return an promise, so await this prmise and no timeout here. Test should be fast and have as less timeouts as possible

Reply:
But here I want to test that none of the users who are not in the channel can get the message.